src/Application.vala

```diff
@@ -96,18 +100,118 @@ public class Screenshot.Application : Gtk.Application {
         if (area) action = 3;
 
         if (action == 0) {
-            if (window == null) {
-                window = new ScreenshotWindow ();
-                window.set_application (this);
-                window.show_all ();
+            take_screenshot.begin ();
+        } else {
+            take_screenshot_backend.begin (action - 1);
+        }
+    }
+
+    private async void take_screenshot () {
+        var portal = new Xdp.Portal ();
+
+        try {
+            var file_uri = yield portal.take_screenshot (null, INTERACTIVE, null);
+
+            pixbuf = new Gdk.Pixbuf.from_file (Filename.from_uri (file_uri, null));
+
+            show_save_dialog ();
+        } catch (Error e) {
+            warning ("Failed to take screenshot via portal: %s", e.message);
+        } finally {
+            release ();
+        }
+    }
+
+    private async void take_screenshot_backend (CaptureType capture_type) {
+        var backend = new ScreenshotBackend ();
+
+        try {
+            pixbuf = yield backend.capture (capture_type, delay, grab_pointer, redact);
+
+            if (pixbuf != null) {
+                show_save_dialog ();
             }
+        } catch (GLib.IOError.CANCELLED e) {
+            //Do nothing
+        } catch (Error e) {
+            show_error_dialog (e.message);
+        } finally {
+            release ();
+        }
+    }
 
-            window.present ();
-        } else {
-            window = new ScreenshotWindow.from_cmd (action, delay, grab_pointer, redact, clipboard);
-            window.set_application (this);
-            window.take_clicked ();
+    private void show_save_dialog () {
```

Comment:
Can we make this take pixbuf as an argument instead of having it as a property?

Reply:
I think this would introduce a memory leak because we would then have to use it as a local variable in the signal handler of `SaveDialog.save_response`